xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -1989,8 +1990,13 @@ public void run() {
     });
     try {
       return future.get(METADATA_SYNC_TIMEOUT_SEC, TimeUnit.SECONDS);
-    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+    } catch (ExecutionException e) {
+      throw new UncheckedExecutionException(e.getCause());
```

Comment:
AFAIK this is a common pattern to unwrap `ExecutionException` because it's designed to wrap underlying exception thrown within a future task.

> Exception thrown when attempting to retrieve the result of a task that aborted by throwing an exception. This exception can be inspected using the Throwable.getCause() method.
> — https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutionException.html

We do this in other places too: https://github.com/grpc/grpc-java/blob/29d238afca83a73ac5c9aeee1539ac1dd7066b0c/stub/src/main/java/io/grpc/stub/ClientCalls.java#L242-L244

(noticed this pattern in the video linked at go/interrupted-exception)

Reply:
FWIW, the example in ClientCalls is passing data to ourselves. The future implementation is at https://github.com/grpc/grpc-java/blob/29d238afca83a73ac5c9aeee1539ac1dd7066b0c/stub/src/main/java/io/grpc/stub/ClientCalls.java#L544

In this specific case, we know the Future implementation is Guava and it is probably fair to trust Guava won't give us garbage and even if it does we just lose some information (vs NPE).

That said, wrapping this in UncheckedExecutionException seems weird, as the caller of the API isn't even aware something is async. Using it anywhere here seems to abuse the exception type. Since this is internal I'd expect it be easiest/best to just wrap the ExecutionException in RuntimeException and move on with life, since we'd never expect this code to run (`future.setException()` is never called). Ditto for CancellationException. TimeoutException and InterruptedException are the cases that actually matter.

Reply:
I'll also note that all this error handling becomes a lot more obvious if you expose the async-ness of the operation to the caller. For public APIs, we may have reason not to do that, but for an internal API with a single caller, there's really no downside.

Reply:
> That said, wrapping this in UncheckedExecutionException seems weird, as the caller of the API isn't even aware something is async

Yes, it is a bit weird. I wasn't sure so much in this decision when chose `UncheckedExecutionException`. I was considering `CompletionException` instead, but it's Java 8: https://github.com/grpc/grpc-java/blob/e0a40c786516c22a83544259b15a3a318f2c6909/xds/src/main/java/io/grpc/xds/XdsClient.java#L501-L504

If I understand it correctly, wrapping `ExecutionException`/`TimeoutException` into `CompletionException` is exactly what `CompletableFuture.join()` is doing:

> better conform with the use of common functional forms, if a computation involved in the completion of this CompletableFuture threw an exception, this method throws an (unchecked) CompletionException with the underlying exception as its cause.
> https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#join--

Compare with [`java.util.concurrent.CompletableFuture#get(long, java.util.concurrent.TimeUnit)`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#get-long-java.util.concurrent.TimeUnit-), which distinguishes between different failure reasons. 

Not sure how join handles `InterruptedException` tho.

Reply:
Ok, apparently `join()` is just not interruptible. It calls `waitingGet(boolean interruptible)` with `interruptible` set to `false`.

Reply:
Refactored to return ListenableFuture instead and shift exception handling decisions to the caller.